rfcs/20231017-collective-broadcast.md

```diff
@@ -0,0 +1,116 @@
+
+## Why call it collective_broadcast and not just broadcast?
+Unfortunately, the op name `broadcast` is already taken by [an op in XLA proper](https://www.tensorflow.org/xla/operation_semantics#broadcast), so we can't have the two names clash. `collective_broadcast` was the preferred alternative.
+
+## Proposed Specification
+
+### collective_broadcast
+
+#### Semantics
+
+Within each process group in the StableHLO process grid, send the value of the
+`operand` tensor from the source process to the target processes and produce a
+`result` tensor. The first process in each of the `replica_groups` is treated as the 
+source process. For any process `n` not in any replica group, that process will just return 
+zeros.
+
+The operation splits the StableHLO process grid into `process_groups` which is
+defined as follows:
+
+* `cross_replica(replica_groups)`
+  if `channel_id <= 0 and use_global_device_ids = false`.
+* `cross_replica_and_partition(replica_groups)`
+  if `channel_id > 0 and use_global_device_ids = false`.
+* `flattened_ids(replica_groups)`
+  if `channel_id > 0 and use_global_device_ids = true`.
+
+Afterwards, within each `process_group`:
+
+* `result@process[result_index] = operand@process[0]`
+
+#### Inputs
+
+| Label | Name                    | Type                                                             | Constraints |
+|-------|-------------------------|------------------------------------------------------------------|-------------|
+| (I1)  | `operand`               | tensor                                                           | (C5)        |
+| (I2)  | `replica_groups`        | variadic number of 1-dimensional tensor constants of type `si64` | (C1-C3)     |
+| (I3)  | `channel_id`            | constant of type `si64`                                          | (C4)        |
+| (I4)  | `use_global_device_ids` | constant of type `i1`                                            | (C4)        |
+
+#### Outputs
+
+| Name     | Type   | Constraints |
+|----------|--------|-------------|
+| `result` | tensor | (C5)        |
+
+#### Constraints
+
+* (C1) `is_unique(replica_groups)`.
+* (C2) `size(replica_groups)` is defined as:
+  * `num_replicas` if `cross_replica` is used.
+  * `num_replicas` if `cross_replica_and_partition` is used.
+  * `num_processes` if `flattened_ids` is used.
+* (C3) `0 <= replica_groups < size(replica_groups)`.
+* (C4) If `use_global_device_ids = true`, then `channel_id > 0`.
+* (C5) `type(result) = type(operand)`.
+
+#### Examples
+
+```mlir
+// num_replicas: 1
+// num_partitions: 4
+// %operand@(0, 0): [[1, 2]]
+// %operand@(0, 1): [[3, 4]]
+// %operand@(0, 2): [[5, 6]]
+// %operand@(0, 3): [[7, 8]]
+%result = "stablehlo.collective_broadcast"(%operand) {
+  replica_groups = dense<[[2, 0], [1, 3]]> : tensor<2x2xi64>,
+  // channel_id = 0
+  channel_handle = #stablehlo.channel_handle<handle = 0, type = 0>
```

Comment:
The channel handle (alias of channel_id) should be 1 for cross_partition communication since we have 4 partitions. Similarly with the next example.
```
channel_handle = #stablehlo.channel_handle<handle = 1, type = 0>
```

Reply:
Done